src/cpu.c

```diff
@@ -440,33 +440,33 @@ static void aggregate(gauge_t *sum_by_state) /* {{{ */
  * current rate; each rate may be NAN. Calculates the percentage of each state
  * and dispatches the metric. */
 static void cpu_commit_one(int cpu_num, /* {{{ */
-                           gauge_t rates[static COLLECTD_CPU_STATE_MAX]) {
+                           gauge_t rates[static COLLECTD_CPU_STATE_MAX],
+                           gauge_t global_rate_sum) {
   metric_family_t fam = {
       .name = "system.cpu.utilization",
+      .help = "Difference in system.cpu.time since the last measurement, "
+              "divided by the elapsed time and number of logical CPUs",
+      .unit = "1",
       .type = METRIC_TYPE_GAUGE,
   };
 
   metric_t m = {0};
-  if (cpu_num == -1) {
-    metric_label_set(&m, "cpu", "total");
-  } else {
+  if (cpu_num != -1) {
     char cpu_num_str[32];
     ssnprintf(cpu_num_str, sizeof(cpu_num_str), "%d", cpu_num);
-    metric_label_set(&m, "cpu", cpu_num_str);
+    metric_label_set(&m, "system.cpu.logical_number", cpu_num_str);
```

Comment:
IMHO using hierarchical dot notation also in metric labels, in addition to the name of the metric itself, is redundant.

Reply:
Yeah, I agree with you. Labels are limited to the metric they are attached to, I don't see the need for introducing namespaces there either. Unfortunately this is what OpenTelemetry has [documented](https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#processor-metrics) as "recommended".

Reply:
At the very top of that document is this statement:
> The system.* namespace SHOULD be exclusively used to report hosts’ metrics. The system.* namespace SHOULD only be used when the metrics are collected from within the target system. (physical servers, virtual machines etc). Metrics collected from technology-specific, well-defined APIs (e.g. Kubelet’s API or container runtimes) should be reported under their respective namespace (e.g. k8s., container.). Resource attributes related to a host, SHOULD be reported under the host.* namespace.

=> `collectd` should **not** be **hard-coding** `system.` prefix to anything

Component knowing in which environment `collectd` is running, can add it instead.

(In Kubernetes / Prometheus case it would be done with `metricRelabelings` / `relabelings` rule in `serviceMonitor` object for `collectd`.)

Reply:
collectd is reading these metrics directly from the operating system, e.g. from `/proc` on Linux. This is exactly what the `system` namespace is for, per the quote you have provided:

> The system.\* namespace SHOULD only be used when the **metrics are collected from within the target system**.

Other metrics, such as the HTTP requests served by Apache, SHOULD NOT go into the `system` namespace.

Reply:
Ah, should have read it closer.  Prefix was about used APIs, rather than whether that info is from host, VM or container.

(Whole /sys, most of /proc, and few devices from /dev, are mounted to all containers by default.)